packages/product/src/services/product-module-service.ts

```diff
@@ -231,6 +231,82 @@ export default class ProductModuleService<
     return [JSON.parse(JSON.stringify(variants)), count]
   }
 
+  async createVariants(
+    data: ProductTypes.CreateProductVariantDTO[],
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<ProductTypes.ProductVariantDTO[]> {
+    const productOptionIds = data
+      .map((pv) => (pv.options || []).map((opt) => opt.option_id!))
+      .flat()
+
+    const productOptions = await this.listOptions(
+      { id: productOptionIds },
+      {},
+      sharedContext
+    )
+
+    const productOptionsMap = new Map<string, ProductTypes.ProductOptionDTO>(
+      productOptions.map((po) => [po.id, po])
+    )
+
+    const productVariantsMap = new Map<
+      string,
+      ProductTypes.CreateProductVariantDTO[]
+    >()
+
+    for (const productVariantData of data) {
+      productVariantData.options = productVariantData.options?.map((option) => {
+        const productOption = productOptionsMap.get(option.option_id!)
+
+        return {
+          option: productOption?.id,
+          value: option.value,
+        }
+      })
+
+      const productVariants = productVariantsMap.get(
+        productVariantData.product_id!
+      )
+
+      if (productVariants) {
+        productVariants.push(productVariantData)
+      } else {
+        productVariantsMap.set(productVariantData.product_id!, [
+          productVariantData,
+        ])
+      }
+    }
```

Comment:
```suggestion
const productVariantsMap = new Map<
  string,
  ProductTypes.CreateProductVariantDTO[]
>()

data
  .map((productVariantData) => {
    const productVariants = productVariantsMap.get(
      productVariantData.product_id!
    ) ?? []
    
    productVariantData.options = (productVariantData.options || []).map((opt) => {
      return {
        option_id: opt.option_id,
        value: opt.value,
      }
    })
    
    productVariants.push(productVariantData)
    productVariantsMap.set(productVariantData.product_id!, productVariants)
  })
```

would this be equivalent?

Reply:
No, it omits existing options and multiple variants having the same product id.